### ISSUE TITLE
Create tests to verify logging in/out works from issue page + home page

### ISSUE THREAD

**Issue by miketaylr** [webcompat/webcompat.com#1385]
So we don't regress again like: https://github.com/webcompat/webcompat.com/issues/1384

We assume logging in and out will work, but don't have any explicit tests for it.

**Comment by miketaylr** [comment]
- [ ] home page
- [ ] /issues route
- [ ] issue page
... more? all routes? dunno.

**Comment by jeanhl** [comment]
I'd like to claim this issue. To clarify:
- testing the login and logout functions in views.py
- Python unittests.

I have a framework in mind but I'd like to discuss it with someone before I go hogwild with it.

**Comment by miketaylr** [comment]
Interesting, I had functional Intern tests in mind since we're doing a GitHub OAuth exchange (hence it not being simple to just mock out). 

What did you have in mind @jeanhl?

**Comment by jeanhl** [comment]
Ah, my bad. I thought I could set up a mock OAuth. I was trying to access github looking at this line in view.py:
"from webcompat import github"
but I can't find a file called "github."

**Comment by miketaylr** [comment]
Ah! 

https://github.com/webcompat/webcompat.com/blob/master/webcompat/__init__.py#L23 is where `github` is defined, it's a reference to the GitHub Flask addon https://github-flask.readthedocs.io/en/latest/.

Mocking it out would be awesome. It would make offline (functional) tests possible.

**Comment by jeanhl** [comment]
I'll look into how to do a mockup of this. The github docs are very useful. Thank you

**Comment by jeanhl** [comment]
I'm looking at the code and I'm thinking of doing the following:
- use flask's test_quest_context to activate a request context temporarily. This will create a temp request, g and session. This should let us test g.referer and request
- use flask's follow_redirects to make sure the redirect works fine

not too sure how to test github.authorize. I get the following when I print github.authorize('public_repo') :
<Response 393 bytes [302 FOUND]>
type: <class 'werkzeug.wrappers.Response'>

Would you like me to check set_referer in helpers.py as well?

If I'm going way off base or if there is a better way, let me know.

**Comment by karlcow** [comment]
+1 to mocking in general. It is something I have been looking at for a longtime, but didn't take the energy to do. 

But before we go off rails 🛤 . Let's make sure we test what we want to test. 

☎️  @miketaylr 
Note that this test is a good thing but would not have detected the issue in #1384 The **login/logout worked**, the issue was with the caching of the resource. Aka basically what should be tested for this #1384 is the following scenario:

1. Go to an `/issues/nnnn`
2. Extract the etag -> etag_before_login
3. Login (or fake login by mocking)
4. Redirect to the `issues/nnnn`
5. Extract the etag -> etag_after_login
6. assert that etag_before_login != etag_after_login

Another way is to check that the status is 200 OK and not 304 Not Modified. This means the resource is coming from the server has changed and so the user sees the template for the login state, this is one possible set of testcases. 

The other one is indeed, make sure the login/logout is working with python unittest. But that's different. And we do that already with intern if I'm right. 

So @miketaylr which one ;)

**Comment by miketaylr** [comment]
You're right @karlcow. In a perfect world we're testing the scenario you describe (to make sure we don't regress #1384), but also have the ability to mock out logging in and out. Either could happen in this bug, and we could open a new bug for the other. 👍

**Comment by brizental** [comment]
I want to work on this! As part of it I'll rewrite the way we mock login, because the way I did it the first time was an accident that worked, lol. I found out that GitHub Flask has mocking already: https://github.com/cenkalti/github-flask/blob/3c60e5876591fda5932d67134be171de17dbd8b6/test_flask_github.py
